### ISSUE TITLE
The exception 'SQLite error: disk is full'  is not submitted to `FileDownloadListener#error`

### ISSUE THREAD

**Issue by metalurgus** [lingochamp/FileDownloader#243]
```
android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount (SQLiteConnection.java)
android.database.sqlite.SQLiteConnection.executeForChangedRowCount (SQLiteConnection.java:731)
android.database.sqlite.SQLiteSession.executeForChangedRowCount (SQLiteSession.java:754)
android.database.sqlite.SQLiteStatement.executeUpdateDelete (SQLiteStatement.java:64)
android.database.sqlite.SQLiteDatabase.updateWithOnConflict (SQLiteDatabase.java:1565)
android.database.sqlite.SQLiteDatabase.update (SQLiteDatabase.java:1514)
com.liulishuo.filedownloader.services.FileDownloadDBHelper.update (SourceFile:340)
com.liulishuo.filedownloader.services.FileDownloadDBHelper.updateError (SourceFile:286)
com.liulishuo.filedownloader.services.FileDownloadRunnable.onError (SourceFile:701)
com.liulishuo.filedownloader.services.FileDownloadRunnable.loop (SourceFile:384)
com.liulishuo.filedownloader.services.FileDownloadRunnable.run (SourceFile:200)
java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1080)
java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:573)
java.lang.Thread.run (Thread.java:838)
```

this is a stacktrace of an error I got from crash reports of firebase. It hurts some of my users. Different Android version (17-23), different devices. 

I see that error themes that user is out of free space. But is that internal system partition, where application data is stored(data/data/<app package name>), or some other partition? And also how can this be fixed, or at least how can I get called back about this error to show some message to user?

**Comment by Jacksgong** [comment]
Sorry to hear that.

---

The filedownloader database is stored in the default database folder. (As Default: `/data/data/<package anme>/databases/`).

## How to fix

Since version 0.3.3, the filedownloader database no longer stores completed tasks data, it means it just stores tasks which unfinished, this will greatly ease the issue you have mentioned.

So, use the latest version, it will solve your issue.

**Comment by metalurgus** [comment]
The version I'm using is exactly 0.3.3:
`compile 'com.liulishuo.filedownloader:library:0.3.3'`

**Comment by Jacksgong** [comment]
OK, In that case, it must be ... as the exception describe "disk is full".  So you need to send a message to the user.

For FileDownloader, in such case, we can't continue to downloading, because we need to store the downloading state to the filedownloader database, which not only is used for resuming tasks from the paused status but also for ensuring FileDownloader can safely resuming tasks from any unexpected status( example: the downloading progress Crash).

**Comment by metalurgus** [comment]
The problem is that there is not callback for this exception, and I can't catch it in my app since it crashes asynchronously in different thread. So to catch it myself, I will anyway have to modify the library code in this case

**Comment by Jacksgong** [comment]
@metalurgus Normally, FileDownloader should handle all exception and carry it to `FileDownloadListener#error` during the entire downloading process.

But in your exception stack trace:

```
...
com.liulishuo.filedownloader.services.FileDownloadDBHelper.updateError (SourceFile:286)
com.liulishuo.filedownloader.services.FileDownloadRunnable.onError (SourceFile:701)
com.liulishuo.filedownloader.services.FileDownloadRunnable.loop (SourceFile:384)
...
```

It must be a bug, what will occur before submitting another error snapshot to `FileDownloadListener#error`, because before submitting an error snapshot to `FileDownloadListener#error` FileDownloader need the downloading status first sync to the filedownloader database.

---

I will fix such bug in next version(v1.0.0)(the exception will be submitted to `FileDownloadListener#error`), thanks for your prompt!

**Comment by Jacksgong** [comment]
Fixed since [0.3.5](https://github.com/lingochamp/FileDownloader/releases/tag/v0.3.5).
